[Kyh196201] 3주차 과제 To-do 리스트 테스트 작성하기

kyh196201 opened the pull request:
트레이너님 안녕하세요!

먼저 List 컴포넌트에 대한 테스트만 작성했습니다.

테스트에 대한 아직 개념이 잘 안잡혀서 조금씩 PR을 올리게될 것 같네요 😂

컴포넌트 별로 어떤 테스트 케이스를 작성해야할지 감이 잘 안오는데 혹시 힌트 좀 주실 수 있으신가요 ?!

kyh196201 committed:
List 컴포넌트 테스트 코드 작성

props로 전달받은 tasks가 빈 배열일 때 텍스트가 나오는지 검증하는 코드를 작성했습니다.

kyh196201 committed:
List 컴포넌트 테스트 케이스 추가 작성

List 컴포넌트가 props로 전달받은 tasks의 내용을 렌더링했는지 체크하는 테스트 케이스를 작성했습니다.

kyh196201 committed:
List 컴포넌트 테스트 코드 추가

tasks props가 빈 배열일 때 toBeInTheDocument을 이용해서
p 엘리먼트가 화면에 렌더링되는지 체크하는 테스트 케이스를 추가했습니다.

kyh196201 committed:
List 컴포넌트 onClickDelete 테스트 코드 작성

props로 전달받은 onClickDelete가 실행되었는지 체크하는 로직을 작성했습니다.

kyh196201 committed:
List 테스트 코드 수정

mock 함수 handleClickDelete를 props로 전달하지 않아서 수정했습니다.

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,63 @@
+});
+
+test('List is Rendered', () => {
+  const tasks = [
+    {
+      id: 1,
+      title: '뭐라도 하기',
+    },
+    {
+      id: 2,
+      title: '아무것도 하지 않기',
+    },
+  ];
```
6번줄이랑 같이 보면, List 컴포넌트의 유즈케이스를 잘 생각하신 것 같아요!👍

kyh196201 replied:
트레이너님 피드백 주신 내용 반영해서 소스 리팩토링 및 추가 작업 진행했습니다.

하나의 테스트가 하나의 역할만 담당하도록 수정했고, 기존에 작성했던 List 컴포넌트 테스트 코드를 BDD 스타일로 개선했습니다.

첨부해주신 링크 보면서 테스트케이스 작성하는 연습을 했는데 너무 많은 도움이 됐습니다 감사합니다 🙌

daadaadaah reviewed (commented):
매우 다양한 쿼리들을 사용해보셨네용! 다양한 시도를 해보시는 거는 너무 좋아요👍
다만, 테스트로 무엇을 기대하는지 의도가 잘 표현이 안된 것 같고, 하나의 테스트에 너무 많은 걸 테스트하는 것 같아요!

BDD 스타일(`describe`-`context`-`it`)로 한번 다시 작성해볼까요?
관련해서 아래 링크들이 도움될 거예요!
[참고]
https://marocchino.net/2016/12/04/about-rspec/
https://ko.javascript.info/testing-mocha#ref-888
https://github.com/ahastudio/til/blob/main/ruby/20161206-rspec-let.md
[Ginkgo - Go 언어 개발자를 위한 BDD 테스팅 프레임워크](https://www.youtube.com/watch?v=gfTsSBRvdqI)

> 컴포넌트 별로 어떤 테스트 케이스를 작성해야할지 감이 잘 안오는데 혹시 힌트 좀 주실 수 있으신가요 ?!

저는 '컴포넌트한테 무엇을 기대하는지'에 대해 고민하며 테스트 케이스를 작성하는 편이예요.
제가 어떻게 테스트 케이스를 발견하는지에 대해 아래 코멘트 참고해보세용😄
https://github.com/CodeSoom/react-week3-assignment-1/pull/134#pullrequestreview-1013694078

kyh196201 committed:
List 컴포넌트 테스트 코드 BDD 스타일로 개선

- 할 일 목록이 없을 경우에 대한 상황과 유즈 케이스를 추가하고, 테스트 코드를 작성했습니다.

kyh196201 committed:
List 컴포넌트 테스트 코드 수정

List 컴포넌트 tasks가 빈 배열일 경우 테스트하는 코드를 리팩토링했습니다.
- render 함수를 실행하는 부분을 it 함수 내부로 옮겼습니다.
- getByText 대신 queryByText를 사용해서 엘리먼트를 검색하도록 수정했습니다.

kyh196201 committed:
List 컴포넌트 테스트 코드 수정

List 컴포넌트가  tasks를 전달받았을 경우 테스트하는 코드를 리팩토링했습니다.
- BDD 스타일로 코드를 개선했습니다.
- 컴포넌트를 렌더링하는 코드가 중복되어서 renderList 함수를 만들었습니다.

kyh196201 commented:
트레이너님 이번에 리팩토링을 진행하면서 하나의 it 함수 내부에 하나의 expect를 실행하도록 코드를 작성했는데,
중복되는 코드가 너무 많아진 것 같습니다.

우선 [leehuun님의 코드](https://github.com/CodeSoom/react-week3-assignment-1/pull/139/files#diff-182f279036de2989c353b641f195e915d226626064775fe367024845ad0a47ceR21)를 참고해서 `renderList`라는 함수를 만들었고, render 함수가 중복되는 부분을 수정했습니다.

혹시 중복되는 코드를 줄일 수 있는 좋은 방법이 있을까요??

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,85 @@
+  ));
+}
+
+const tasks = [
+  {
+    id: 1,
+    title: '뭐라도 하기',
+  },
+  {
+    id: 2,
+    title: '아무것도 하지 않기',
+  },
+];
```
여러 곳에서 사용되는 tasks 데스트용 데이터를 보통 fixture라고 이야기합니다.
저는 이러한 데이터를 `fixtures` 폴더 하나 만들어서 따로 관리하는 편이예요!

kyh196201 replied:
따로 폴더 만들어서 관리하면 훨씬 편하겠네요!!
좋은 방법 알려주셔서 감사합니당 🙌

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,85 @@
+      onClickDelete={handleClickDelete}
+    />
+  ));
+}
```
중복되는 코드를 함수로 추출하신 거 너무 잘하셨어요!👍
테스트 코드도 실제 구현 코드와 마찬가지로 리팩터링이 필요합니다!

kyh196201 replied:
감사합니다 😀
나머지 컴포넌트 테스트 코드 작성하면서  리팩터링 진행하도록 하겠습니다 ㅎㅎ!

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,85 @@
+  });
+
+  context('할 일 목록이 있으면', () => {
+    it('할 일 개수만큼 Item 컴포넌트가 보입니다', () => {
+      const { queryAllByRole } = renderList({
+        tasks,
+      });
+
+      expect(queryAllByRole('listitem')).toHaveLength(tasks.length);
+    });
+
+    it('첫번째 Item 컴포넌트의 내용이 "뭐라도 하기"입니다', () => {
+      const { queryAllByRole } = renderList({
+        tasks,
+      });
+
+      const listItems = queryAllByRole('listitem');
+
+      expect(listItems[0]).toHaveTextContent(tasks[0].title);
+    });
```
저라면, '할일들이 보인다' 라고 1개의 테스트로 통합시키고, 반복문으로 expect를 돌릴 것 같아요!
아, 혹시, `it` 하나에 `expect` 1개로 하고 싶어서 이렇게 하셨나용?

kyh196201 replied:
넵 맞습니다! `it` 하나에 하나의 `expect`만 있는게 좋다고 하셔서 이렇게 작성했는데, `tasks`를 반복문으로 돌리면서 `expect`를 실행해도 될까요??

daadaadaah reviewed (commented):
1. 어느 곳에서는 get*를 사용하셨고 어느 곳에서는 query* 사용하셨는데, 혹시 이 둘의 차이점에 대해 알고 계신가요? 언제 get*을 사용하는게 좋은지, 언제 query*를 사용해야 하는지 공부해보셨으면 좋겠습니다!
도움될만한 링크 하나 남겨요
참고 : https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#using-query-variants-for-anything-except-checking-for-non-existence

> 혹시 테스트 코드를 작성할 때 중복되는 코드를 줄일 수 있는 좋은 방법이 있을까요??

말씀하시는 중복코드가 renderList로 빼신 그 중복코드 말씀하시는건가용?
맞다면, 이 중복코드는 어쩔 수 없이 생기는것 같아요! 그래서, 테스트 코드도 리팩터링이 필요하죠!
지금 하신 것처럼 함수로 빼서 중복된 코드를 관리하는 거 너무 잘하신거예요!👍

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,85 @@
+    });
+  });
+
+  context('사용자가 완료 버튼을 클릭하면', () => {
+    const handleClickDelete = jest.fn();
+
+    it('onClickDelete 함수가 실행됩니다.', () => {
+      const { getAllByRole } = renderList({
+        tasks,
+        handleClickDelete,
+      });
+
+      const completeButtons = getAllByRole('button');
+
+      fireEvent.click(completeButtons[0]);
+
+      expect(handleClickDelete).toBeCalled();
+    });
+
+    it('onClickDelete의 인자로 클릭한 Item의 id가 전달됩니다.', () => {
+      const { getAllByRole } = renderList({
+        tasks,
+        handleClickDelete,
+      });
+
+      const completeButtons = getAllByRole('button');
+
+      fireEvent.click(completeButtons[0]);
+
+      expect(handleClickDelete).toBeCalledWith(1);
+    });
+  });
```
저라면 "할일 완료 버튼 클릭" 이라는 테스트 대상으로 보고 테스트를 작성할 것 같아요!
그래서, `context`보다 `describe`로 표현할 것 같아요! 
```js
describe('완료 버튼 클릭', () => {
  it('handleClickDelete 이 호출된다.', () => {
  });
});
```

kyh196201 replied:
아직 describe, context, it의 역할이 헷갈려서 계속 참고하면서 작성했습니다 감사합니다 ㅎㅎ

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,85 @@
+  context('사용자가 완료 버튼을 클릭하면', () => {
+    const handleClickDelete = jest.fn();
+
+    it('onClickDelete 함수가 실행됩니다.', () => {
```
handleClickDelete 함수가 호출되는 걸로 테스트를 작성하셔서 
'handleClickDelete 함수가 호출된다'라고 기재하는게 맞는것 같아요!

kyh196201 replied:
앗 바로 수정해야겠어요 감사합니다 ㅎㅎ

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,85 @@
+        handleClickDelete,
+      });
+
+      const completeButtons = getAllByRole('button');
```
button으로 요소를 찾는 것보다 "완료"라는 의미로 찾을 수 있게 query를 수정해보시는건 어떤가요?
button은 추가버튼으로 볼 수 있고, 수정버튼도 될 수 있고 해서요!

kyh196201 replied:
피드백 감사합니다 트레이너님!
이 부분 코드 작성하면서 많이 고민됐는데 "완료" 텍스트로 찾으면 됐었네요 수정하겠습니다 😀

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,85 @@
+      expect(handleClickDelete).toBeCalled();
+    });
+
+    it('onClickDelete의 인자로 클릭한 Item의 id가 전달됩니다.', () => {
+      const { getAllByRole } = renderList({
+        tasks,
+        handleClickDelete,
+      });
+
+      const completeButtons = getAllByRole('button');
+
+      fireEvent.click(completeButtons[0]);
+
+      expect(handleClickDelete).toBeCalledWith(1);
+    });
```
이 테스트는 불필요한 테스트 같아요!ㅎㅎ

kyh196201 replied:
이 부분은 `Item` 컴포넌트에서 테스트할 수 있는 내용이라 불필요한 테스트인건가요??

daadaadaah replied:
아니요! 테스트 설명은 다르게 기재하셨는데, 테스트 코드 작성해두신거 보면, 이미 바로 위에서 한 테스트랑 겹쳐서요!
그리고, "id가 잘 전달 되는지 여부"는 저라면 테스트 안할 것 같아요! 전달 여부 보다 전달되어서 어떤 일이 벌어졌는지가 더 중요해서 이 부분에 대해 테스트 할 것 같은데, 이 테스트가 바로 위에 한 테스트라서요!

kyh196201 committed:
List 컴포넌트 리팩터링

- 사용한 쿼리를 queryBy에서 getBy로 변경했습니다.
- 완료 버튼 클릭을 테스트하는 부분을 context에서 describe로 변경했습니다
- handleClickDelete에 인자가 전달되었는지 테스트하는 코드를 삭제했습니다.

kyh196201 committed:
List.test.jsx 리팩터링

- 내부적으로 사용하는 tasks를 fixtures/tasks.js 파일 내에서 따로 선언해서 사용했습니다.
- tasks의 title이 렌더링되었는지 확인하는 테스트 코드를 반복문으로 개선했습니다.

kyh196201 committed:
Input 컴포넌트 테스트 코드 추가

kyh196201 committed:
Page 컴포넌트 테스트 코드 작성

Item 컴포넌트와 List 컴포넌트가 렌더링되었는지 확인하는 테스트 케이스만 작성했습니다.

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,63 @@
+      expect(getAllByRole('listitem')).toHaveLength(TASKS.length);
+    });
+
+    TASKS.forEach(({ title }, index) => {
+      it(`${index + 1}번째 Item 컴포넌트의 내용인 "${title}"가 보인다.`, () => {
+        const { container } = renderList({
+          tasks: TASKS,
+        });
+
+        expect(container).toHaveTextContent(title);
+      });
```
kyh196201님처럼 해도 되고, 아래처럼 해도 될 것 같아요!
```jsx
    it('할일들이 보인다', () => {
      TASKS.forEach(({ title }, index) => {
          expect(container).toHaveTextContent(title);
      });
    });
```
지금 생각해보니, 제가 1개의 it에 1개의 expect를 사용한다는 이야기를 좀더 자세히 설명해줬으면 좋았다는 생각이 드네요.
이 의미가 `1개의 테스트에 1개만 기대하자`라는 의미였어요.
나쁜 예로 it 안에 input 존재여부, 버튼 호출여부 등 여러개가 하나의 it에 들어간 경우가 있고,
반면, 지금처럼 복수형 아이템들(예 : `할일들` 또는 `상품들`)을 보여줄 때,  위 코드처럼 표현하기도합니다!

제가 이렇게 하는걸 선호한다는 것이지 이게 꼭 정답이 아니라는 점 항상 명심해두셨으면 좋겠어요!
저는 describe-context-it 패턴을 선호하지만, 어떤 개발자는 test 라는 것 안에 모든 테스트를 다 넣을 수도 있거든요!

PR도 열심히 질문도 열심히 해주는 모습 너무 좋습니다👍 남은 기간에도 같이 화이팅해봐용!!🚀

kyh196201 replied:
트레이너님 상세한 피드백 너무 감사합니다! 🙂

하나의 it이 하나의 관심사에 대한 테스트를 담당해야 된다는 의미로 이해했습니다! ㅎㅎ

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,63 @@
+
+const handleChangeTitle = jest.fn();
+const handleClickDeleteTask = jest.fn();
+const handleClickAddTask = jest.fn();
```
아예 밖으로 빼신 이유가 있으신가요?

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,63 @@
+  ));
+}
+
+beforeEach(() => {
+  jest.clearAllMocks();
+});
```
아예 밖으로 빼신 이유가 있으신가요?
보통 describe 내에 선언해주거든요!

kyh196201 replied:
앗 전역으로 사용하고싶어서  `mock` 함수와 `beforeEach`를 제일 상단에 작성했었습니다.
`describe` 함수 내부로 변경하겠습니다 감사합니다!

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,63 @@
+  it('Input이 보인다.', () => {
+    const { getByLabelText } = renderPage();
+
+    const input = getByLabelText('할 일');
+
+    expect(input).toBeInTheDocument();
```
지금은 input 변수화가 필요 없어보여요! 그래서 아래처럼 작성해주는 것이 좋을 것 같아요! 
```jsx
 expect(getByLabelText('할 일')).toBeInTheDocument();
```

kyh196201 replied:
불필요한 변수화 제거했습니다 피드백 감사합니다 🙌

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,63 @@
+        tasks: TASKS,
+      });
+
+      TASKS.forEach((task) => {
+        expect(container).toHaveTextContent(task.title);
+      });
```
오! 여기서 사용하셨네용! 잘하셨어요!👍

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,63 @@
+      });
+
+      expect(queryByText('할 일이 없어요!')).not.toBeInTheDocument();
+    });
```
이거까지 고려한건 좋았지만, 위에 할일 들이 보인다라는 테스트가 있으므로, 삭제해도 좋을 것 같아요!

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,74 @@
+    expect(getByText('추가')).toBeInTheDocument();
+  });
+
+  it('input의 현재 값과 props.value의 값이 서로 같다.', () => {
+    const taskTitle = '할일';
+
+    const { getByPlaceholderText } = renderInput(taskTitle);
+
+    expect(getByPlaceholderText('할 일을 입력해 주세요')).toHaveValue(taskTitle);
+  });
```
정말 경우의 수에 대해 촘촘히 생각하셨네용👍

이부분 까지 테스트 하고 싶다고 한다면,

저라면, 24줄~31줄이랑 해서 `context`로 분기할 것 같아요!
예를 들어, 1) 입력값이 없다면, 2) 입력값이 있다면, 이런 식으로요!

kyh196201 replied:
props로 전달받은 value가 화면에 렌더링되었는지 확인하고 싶어서 테스트 코드를 작성했는데, `context`와 같이 작성했으면 더 좋았을 것 같네요 !!
피드백 감사합니다 🙌

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,74 @@
+
+import Input from './Input';
+
+const handleChangeTitle = jest.fn();
+const handleClickAddTask = jest.fn();
```
아예 밖으로 빼신 이유가 있으신가요?
혹시, renderInput 함수 때문인가용?

kyh196201 replied:
네 맞습니다!

전체 테스트 코드에서 사용하고, `renderInput` 내부에서 참조하기 위해서 전역으로 선언했습니다.

위 리뷰에서 보통 `describe` 함수 내부에서 선언한다고 말씀해주셨는데, 전역 변수 사용을 지양하기 위해서 그런건가요??

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,74 @@
+  ));
+}
+
+beforeEach(() => {
+  handleChangeTitle.mockClear();
+  handleClickAddTask.mockClear();
+});
```
아예 밖으로 빼신 이유가 있으신가요?

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,74 @@
+    const input = getByPlaceholderText('할 일을 입력해 주세요');
+
+    expect(input).toBeInTheDocument();
+    expect(input).toHaveAttribute('type', 'text');
```
내부 구현사항과 관련된 테스트라서 저라면 뺄 것 같아요!

kyh196201 replied:
테스트의 범위?를 정하는게 아직 어려운 것 같습니다 😂

트레이너님은 이 부분이 컴포넌트의 역할 혹은 동작과는 직접적인 관련이 없는 부분이라 빼시는 건가요??

daadaadaah reviewed (commented):
테스트 촘촘히 잘 작성해주신 것 같아요!👍 테스트 코드 작성에 대해 거의 감을 잡으신것 같구!👏👏👏👏👏👏

조금더 익숙해지시면, 테스트 설명도 확장성 있게 작성해보시는거 도전해보세요! 
지금은 테스트 코드를 많이 작성해보는 것을 포커스로 연습하시구용!😎


> 1. 이번 과제를 하면서 `Item > List > Page` 이렇게 하위 컴포넌트 먼저 테스트 코드를 작성했는데, 실무에서 테스트 코드를 작성할 때도 하위 컴포넌트 먼저 테스트 코드를 작성하나요??

상황마다 달라요! 예를 들어, 회원가입 페이지 만든다고 하면, 상위 컴포넌트부터 테스트 코드를 작성할 것이고, 만약, 이미 회원가입 페이지가 있는데, 거기에 이메일 입력란을 추가하라고 하면, 작은 하위컴포넌트인 이메일 입력란 테스트 코드를 먼적 작성하는 편입니다! 어떤것을 먼저 작성해야 된다는 답은 없어요! 상향식이랑 하향식 모두 왔다갔다 할 줄 알아야 합니다!ㅎㅎ


> 2. `Page 컴포넌트`의 테스트 케이스를 작성할 때 `List 컴포넌트`의 테스트 케이스와 중복 되는 내용이 생기더라구요, 상위 컴포넌트의 경우 하위 컴포넌트의 테스트 내용을 모두 포함해서 작성해야 하나요???

저는 모두 작성하는 편이예요! 왜냐하면, List 컴포넌트만 따로 테스트할 때 동작하던 것이 Page 내에 넣었을 때 동작하지 않을 수 있다는 생각 때문에, 둘다 작성하는 편입니다! 테스트 작성 범위를 어느정도로 할 것이냐는 정답이 없는 것 같아요!

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,63 @@
+
+  context('할 일 목록이 없으면', () => {
+    it('"할 일이 없어요!" 텍스트가 보인다.', () => {
+      const { container } = renderPage();
```
저는 "할일 없다"는 의도를 드러내주기 위해 아래처럼 작성하기도 합니다!
```jsx
   const { container } = renderPage({
        tasks:[],
      });

```

kyh196201 committed:
Page 컴포넌트 테스트 코드 리팩토링

- 불필요한 input 변수를 제거했습니다.
- mock 함수와 beforeEach를 describe 내부에서 선언했습니다.
- 할 일 목록이 있을 경우 "할 일이 없어요!" 텍스트가 노출되는지 확인하는 테스트 코드가 불필요하다고 생각하여 제거했습니다.

kyh196201 committed:
Input 컴포넌트 테스트 코드 리팩토링

- input 태그의 내부 구현 사항을 검증하는 테스트 코드를 제거했습니다.
- 입력 값과 input 태그의 value가 같은지 비교하는 테스트 코드를 context 내부로 옮겼습니다.

kyh196201 committed:
Input 컴포넌트 테스트 코드 리팩토링

- mock  함수와 beforeEach 함수를 describe 내부로 이동했습니다.

kyh196201 committed:
Page 컴포넌트 테스트 코드 추가 구현

- input 엘리먼트를 찾는데 사용되는 '할 일' 텍스트를 LABEL_TEXT로 변수화했습니다.
- props로 전달받은 이벤트 핸들러의 동작을 확인하는 테스트 코드를 추가했습니다.

kyh196201 commented:
트레이너님 피드백 해주신 내용 반영했고, Page, App 컴포넌트 테스트 코드 추가 작성했습니다.
피드백 부탁드립니다 🙂

그리고 질문이 있습니다!

---

1. 테스트 제목 작성에 대해서

`it`을 작성할 때 테스트 제목을 어떻게 작성할지 고민이 되더라구요.

만약 아래와 같은 케이스가 있다면 어떤게 더 좋은 테스트 제목일지 궁금합니다.

1번이 좀더 구체적이고, 2번이 좀더 추상적인 느낌이라고 생각되는데 이 부분이 계속 고민되더라구요..!

```
1. props.tasks가 빈 배열이면

2. 할 일 목록이 없으면
```

2. App 컴포넌트 테스트 코드에 대해서

이번에 App 컴포넌트에 대한 테스트 코드를 작성하지 못했습니다.
제 생각에 App 컴포넌트는 `Input`, `Item`에서 이벤트가 실행되면 실제로 이벤트가 실행되었고, 내부 state가 변경되었는지를 테스트해야될 것 같은데 맞을까요??

kyh196201 committed:
App 컴포넌트 테스트 코드 추가

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,75 @@
+    expect(getByText('추가')).toBeInTheDocument();
+  });
+
+  context('입력 값이 있다면', () => {
+    it('input의 현재 값과 props.value의 값이 서로 같다.', () => {
+      const taskTitle = '할일';
+
+      const { getByPlaceholderText } = renderInput(taskTitle);
+
+      expect(getByPlaceholderText('할 일을 입력해 주세요')).toHaveValue(taskTitle);
+    });
+  });
```
"입력값이 없다면" 이라는 context가 있다면, 필요한 테스트일 것 같지만, 지금 작성해놓으신 코드를 보면, 생략해도 되는 테스트인것 같아요!

kyh196201 replied:
피드백 감사합니다!
context 부분 제거하고 it에 있는 내용도 변경했습니다🙌

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,63 @@
+
+import TASKS from './fixtures/tasks';
+
+function renderList({ tasks = [], handleClickDelete = () => {} }) {
+  return render((
+    <List
+      tasks={tasks}
+      onClickDelete={handleClickDelete}
+    />
+  ));
+}
```
이것도 describe 안에 넣읍시다!

kyh196201 replied:
앗 이 부분 빼먹었네요 수정하겠습니다!

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,63 @@
+      });
+
+      expect(getAllByRole('listitem')).toHaveLength(TASKS.length);
+    });
```
바로 밑에 테스트가 있으므로, 생략해도 될 것 같아요!

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,109 @@
+    expect(getByLabelText(LABEL_TEXT)).toBeInTheDocument();
+  });
+
+  context('taskTitle이 있으면', () => {
+    it('input의 value가 taskTitle과 같다.', () => {
+      const taskTitle = '입력값';
+
+      const { getByLabelText } = renderPage({
+        taskTitle,
+      });
+
+      expect(getByLabelText(LABEL_TEXT)).toHaveValue(taskTitle);
+    });
+  });
```
"taskTitle이 없다면" 이라는 context가 있다면, 필요한 테스트일 것 같지만, 지금 작성해놓으신 코드를 보면, 생략해도 되는 테스트인것 같아요!

kyh196201 replied:
`context`가 하나만 사용되는 부분 모두 수정하겠습니다 피드백 감사합니다 😊

daadaadaah reviewed (commented):
아주 잘 하고 있어요! 👍 

renderXXX 함수 만들때, RORO 패턴 한번 사용해보세요!

그리고, 저는 테스트코드 작성할 때는 renderXXX 같은 함수 내의 매개변수 초기값을 따로 설정해주지 않고, 

```jsx
  function renderXXXX({todos = [] }) {
    return render(    
    <XXX
     // ..props ~
    />);
  }

```
아래 코드처럼 renderXXX() 를 사용하는 곳에서 하나하나 명시해주는 편이예요!
이렇게 하면, 좀더 테스트 의도를 더 드러낼수 있더라구요!

```jsx
  function renderXXXX({todos}) {
    return render(    
    <XXX
     // ..props ~
    />);
  }

context('할일이 없다면', () => {

    it('할일 없어요가 보인다', () => {
            renderXXX({ todos: []});
     });
});
```



> 1. 테스트 제목 작성에 대해서
`it`을 작성할 때 테스트 제목을 어떻게 작성할지 고민이 되더라구요.
만약 아래와 같은 케이스가 있다면 어떤게 더 좋은 테스트 제목일지 궁금합니다.
1번이 좀더 구체적이고, 2번이 좀더 추상적인 느낌이라고 생각되는데 이 부분이 계속 고민되더라구요..!
> ```
> 1. props.tasks가 빈 배열이면
> 
> 2. 할 일 목록이 없으면
> ```

2번이라고 생각해요! 1번은 구현에 포커스로 두고 작성한 것 같고, 2번은 사용자 입장에 포커스로 두고 작성한 것 같아서요! 그리고, 테스트 제목이 너무 구체적이면 구현이 바뀌었을 때, 테스트 제목도 수정이 필요합니다. 그래서, 2번처럼 해놓으면, props.tasks -> props.todos로 바뀌어도 테스트 제목을 수정할 필요가 없게될 것 같아요!


> 2. App 컴포넌트 테스트 코드에 대해서
> 이번에 App 컴포넌트에 대한 테스트 코드를 작성하지 못했습니다.
제 생각에 App 컴포넌트는 `Input`, `Item`에서 이벤트가 실행되면 실제로 이벤트가 실행되었고, 내부 state가 변경되었는지를 테스트해야될 것 같은데 맞을까요??

내부 state가 변경되었는지의 여부가 예를 들어, 할일을 추가했을 때, "할일이 보인다"라는 이야기라면, 맞아요!👍

kyh196201 committed:
List 컴포넌트 테스트 코드 리팩터링

- mock 함수의 위치를 describe 함수 내부로 이동했습니다.
- 할 일 목록이 렌더링 되었는지 테스트하는 코드에서 중복되는 내용을 제거했습니다.

kyh196201 committed:
Page 컴포넌트 테스트 코드 리팩터링

- 불필요한 context 문을 제거했습니다.
- 테스트의 의도를 드러내기 위해서 renderPage 함수에 빈 tasks를 파라미터로 전달했습니다.

kyh196201 committed:
Input 컴포넌트 테스트 코드 리팩터링

- 불필요한 context 문을 제거했습니다.
- 입력 중인 값이 있는지 테스트하는 코드의 설명을 구현 보다는 유즈 케이스에 적합한 내용으로 수정했습니다.

daadaadaah reviewed (commented):
진짜 잘하셨어요!👍 App 컴포넌트도 테스트 작성 다 완료하신건가요?

kyh196201 commented:
트레이너님 App 컴포넌트에 대한 테스트 코드는 아직 작성하지 못했습니다.
과제 풀이와 다른 분들의 코드를 보면서 작성하려고 합니다.

이번주는 과제를 완료하지 못했네요,,, 다음주는 열심히해서 과제 미리 완료하도록 하겠습니다!
 항상 좋은 피드백 주셔서 감사합니다 😊

kyh196201 committed:
App 컴포넌트 테스트 코드 추가

- input에 대한 테스트 코드를 작성했습니다.
- 할 일 목록이 있을 경우와 없을 경우로 상황을 나눠서 테스트 코드를 작성했습니다.
- 각 이벤트가 실행되었을 경우에 대한 테스트 코드를 작성했습니다.
